windows.go

```diff
@@ -37,17 +38,29 @@ func NewWatcher() (*Watcher, error) {
 		return nil, os.NewSyscallError("CreateIoCompletionPort", e)
 	}
 	w := &Watcher{
-		port:    port,
-		watches: make(watchMap),
-		input:   make(chan *input, 1),
-		Events:  make(chan Event, 50),
-		Errors:  make(chan error),
-		quit:    make(chan chan<- error, 1),
+		port:      port,
+		watches:   make(watchMap),
+		input:     make(chan *input, 1),
+		Events:    make(chan Event, 50),
+		Errors:    make(chan error),
+		quit:      make(chan chan<- error, 1),
+		recursive: false,
 	}
 	go w.readEvents()
 	return w, nil
 }
 
+// SetRecursive makes future Add() calls watch directories recursively
+// if this file system supports that.
+// Must be called before Add/Remove.
+func (w *Watcher) SetRecursive() error {
```

Comment:
I'm not sure about this API -- what do others think?

Adding a separate API is good in that there aren't any breaking changes for existing users.

There is no way to disable the recursive flag. Is that important?

Reply:
ya, i could go either way on this.

my thinking was that you want the API to be as restrictive as possible - no switching back and forth between recursive and non-recursive, no switching modes once the watcher has started listening on events.

But certainly `SetRecursive(bool)` would be a lot more idiomatic.

Reply:
There are a lot of different options, and it's important to get the API right, because the same API would need to support a user-space recursive watcher (#18) and [FSEvents](https://github.com/fsnotify/fsevents) on macOS.

@pbnjay @rogpeppe Having look at the Mac side, do you have any opinions on how this API should work?

If I recall, macOS is always recursive when making a watch?

This option seems pretty good and is backwards compatible -- a method on the watcher.

Other options would be some other form of NewWatcher with configuration. If we want the watcher to be 100% recursive or 100% not, this may be better.

Or having a separate `AddRecursive` function, which would allow intermixing of recursive and not-recursive. But we may wind up with too many variants -- like with and without Symlinks, Recursive, Ops filtering, etc.

Reply:
@nathany 

I can confirm that fsevents on macOS is recursive by default.

I am in support of `Other options would be some other form of NewWatcher with configuration. If we want the watcher to be 100% recursive or 100% not, this may be better`.

I don't see a concrete use case in which we want to keep switching back and forth. Most of time, people want a recursive watch. If there are paths that they are not interested in, they can reasonably add additional filtering-out step at their application level to ignore such events.

Please let us know what else to do for this to be acceptable. thanks

Reply:
As a side note, we forked this and enabled a recursive watch like this in our production software and we don't observe issues https://github.com/fsnotify/fsnotify/issues/259 and https://github.com/fsnotify/fsnotify/issues/243.

Reply:
I think we could just go with this API for now. 

It's possible to change it later in a v2 of the library.

We probably just need someone to update the branch, merge it, and do the release

Reply:
(Sorry this somehow skipped my inbox)

Yes macOS is recursive by default, but it could be a challenge to implement logic to "disable" that however. It sounds like could be the default fsnotify semantics?

Reply:
One possibility would be to use `Add("dir/**")` or `Add("dir/...")`.

Especially since recursive operations may be expensive on some platforms, there may be use cases where you might want things like "recursively watch ~/.cache/go-build, and non-recursively watch ~/.cache/go-mod", or something like that.

The advantage of using a string like that is that #352 can be added without too much friction too as e.g. `AddMask("path", fsnotify.CREATE | fsnotify.DELETE)` or some such.

Reply:
Would it make sense to have a syntax like Glob?

https://pkg.go.dev/path/filepath@go1.18.4#Glob
https://pkg.go.dev/io/fs@go1.18.4#Glob

Reply:
One challenge with this approach is what happens on BSD (kqueue) or Linux (inotify), since only Windows and FSEvents on macOS support recursive watching. 

Is it an error on some platforms? At least until such time as a user-space recursive watcher (#18) covers those platforms?

Reply:
> One challenge with this approach is what happens on BSD (kqueue) or Linux (inotify), since only Windows and FSEvents on macOS support recursive watching.

Yeah, .Add() can return an error in those cases, similar to how SetRecursive() does now. Something like:


	func recursive(path string) (string, bool) {
		if strings.HasSuffix(path, "/...") || strings.HasSuffix(path, `\...`) {
			return path[:len(path)-4], true
		}
		return path, false
	}

	path, recurse := recursive(path)
	if recurse {
    	return errors.New("recursion not supported on this system")
	}

Or for Windows and other systems where it works:

    recurse := false
    path, recurse = recursive(path)


> Would it make sense to have a syntax like Glob?

I'd be in favour of that; there's an old issue for that (#10).

Using the `/...` syntax is probably better, since e.g. something like `**/*.zip` won't work and I don't think we need to support that (it'll be tricky to get that to work well for all cases).

Reply:
Yeah, that makes sense, and there is precedence for that format (e.g. `go test ./...`). Sounds good.